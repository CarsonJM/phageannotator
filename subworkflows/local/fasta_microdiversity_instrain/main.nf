//
// Assess virus microdiversity with instrain
//

include { GUNZIP as GUNZIP_FASTA    } from '../../../modules/nf-core/gunzip/main'
include { GUNZIP as GUNZIP_PROTEINS } from '../../../modules/nf-core/gunzip/main'
include { INSTRAIN_PROFILE          } from '../../../modules/nf-core/instrain/profile/main'
include { INSTRAIN_COMPARE          } from '../../../modules/nf-core/instrain/compare/main'


workflow FASTA_MICRODIVERSITY_INSTRAIN {
    take:
    bam             // [ [ meta ], bam ]        , BAM files from reads aligned to FASTA file (mandatory)
    fasta_gz        // [ [ meta ], fasta.gz ]   , FASTA file used in read alignment (mandatory)
    proteins_fna_gz // [ [ meta ], fna.gz ]     , FASTA file for protein-coding genes (optional)
    instrain_stb    // [ [ meta ], stb.tsv ]    , TSV file with two columns for associationg scaffolds to bins (optional)

    main:
    ch_versions = Channel.empty()

    //
    // MODULE: gunzip FASTA file
    //
    fasta = GUNZIP_FASTA ( fasta_gz ).gunzip

    //
    // MODULE: gunzip proteins fna file
    //
    proteins_fna = GUNZIP_PROTEINS ( proteins_fna_gz ).gunzip
    ch_versions = ch_versions.mix(GUNZIP_PROTEINS.out.versions)

    // remove meta information for instrain profile inputs
    ch_stb_file_tsv_nometa = instrain_stb.map { it[1] }.first()
    ch_fasta_nometa = fasta.map { it[1] }.first()
    ch_proteins_fna_nometa = proteins_fna.map { it[1] }.first()

    //
    // MODULE: Profile microdiveristy within each sample
    //
    ch_instrain_profiles = INSTRAIN_PROFILE ( bam, ch_fasta_nometa, ch_proteins_fna_nometa, ch_stb_file_tsv_nometa ).profile
    ch_instrain_gene_tsv = INSTRAIN_PROFILE.out.gene_info
    ch_versions = ch_versions.mix(INSTRAIN_PROFILE.out.versions)

    // combine bams and profiles across samples
    ch_bam_profiles_combined = bam.join(ch_instrain_profiles)
    ch_instrain_compare_input = ch_bam_profiles_combined.map { [ [ id:'all_samples' ], it[1], it[2] ] }.groupTuple( sort: 'deep' )

    //
    // MODULE: Compare microdiversity across samples (within groups OR across all samples)
    //
    ch_instrain_compare = INSTRAIN_COMPARE ( ch_instrain_compare_input, ch_stb_file_tsv_nometa ).compare
    ch_instrain_comparisons_tsv = INSTRAIN_COMPARE.out.comparisons_table
    ch_instrain_pooled_snvs_tsv = INSTRAIN_COMPARE.out.pooled_snv

    ch_versions = ch_versions.mix(INSTRAIN_COMPARE.out.versions)

    emit:
    gene_info_tsv   = ch_instrain_gene_tsv  // [ [ meta ], gene_info.tsv ]
    versions        = ch_versions           // [ versions.yml ]
}
