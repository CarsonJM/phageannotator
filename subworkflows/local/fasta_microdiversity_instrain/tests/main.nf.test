nextflow_workflow {

    name "Test Subworkflow: FASTA_MICRODIVERSITY_INSTRAIN"
    script "../main.nf"
    workflow "FASTA_MICRODIVERSITY_INSTRAIN"
    tag "subworkflows"
    tag "subworkflows_local"
    tag "fasta_microdiversity_instrain"
    tag "fasta_microdiversity_instrain_default"
    tag "instrain/compare"
    tag "instrain/profile"
    tag "gunzip"

    test("Parameters: default") {

        when {
            params {
                outdir              = "$outputDir"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [ [ id:'test' ], file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExists: true) ],
                    [ [ id:'test2' ], file(params.test_data['homo_sapiens']['illumina']['test2_paired_end_sorted_bam'], checkIfExists: true) ]
                )
                input[1] = Channel.of([ [ id: 'reference_fasta' ], file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true) ])
                input[2] = null
                input[3] = Channel.of([ [ id: 'instrain_stb' ], file(params.test_data['modules_nfcore']['instrain_stb'], checkIfExists: true) ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
