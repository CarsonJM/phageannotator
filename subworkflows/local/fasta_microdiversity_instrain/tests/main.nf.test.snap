{
    "bam & fasta & stb_file": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    "versions.yml:md5,4f3be9f56f69dcce1903598ad2fc9635",
                    "versions.yml:md5,4f3be9f56f69dcce1903598ad2fc9635",
                    "versions.yml:md5,8ab3c0c738c13fcd51c4c190273bf109"
                ],
                "gene_info_tsv": [
                    
                ],
                "versions": [
                    "versions.yml:md5,4f3be9f56f69dcce1903598ad2fc9635",
                    "versions.yml:md5,4f3be9f56f69dcce1903598ad2fc9635",
                    "versions.yml:md5,8ab3c0c738c13fcd51c4c190273bf109"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-19T17:05:29.756590992"
    }
}